logrotate.c

```diff
@@ -2909,6 +2886,51 @@ static int readState(const char *stateFilename)
     return 0;
 }
 
+static int lockState(const char *stateFilename)
+{
+    int lockFd = open(stateFilename, O_RDWR | O_CLOEXEC);
```

Comment:
With `debug == 1` we either need to open the state file with `O_RDONLY`, or rather skip the locking entirely.  Otherwise we break the use case where a non-privileged runs `logrotate -d /etc/logrotate.conf`:

With current master (7da0c415da2ad1f56b67cbdf7ad159a9fe5e6d2a)
```
$ ./logrotate -d /etc/logrotate.conf
WARNING: logrotate in debug mode does nothing except printing debug messages!  Consider using verbose mode (-v) instead if this is not what you want.

reading config file /etc/logrotate.conf
including /etc/logrotate.d
Ignoring yum.rpmsave, because of *.rpmsave pattern match
reading config file bootlog
reading config file btmp
reading config file chrony
reading config file cyrus-imapd
reading config file dnf
reading config file dpkg
reading config file fedmsg
reading config file glusterfs
reading config file httpd
reading config file iscsiuiolog
reading config file libvirtd
reading config file libvirtd.qemu
reading config file numad
reading config file ppp
reading config file squid
reading config file tinyproxy
reading config file vsftpd
reading config file wtmp
reading config file yum
Reading state from file: /var/lib/logrotate.status
Allocating hash table for state file, size 64 entries
[...]
```

With this pull request:
```
$ ./logrotate -d /etc/logrotate.conf
WARNING: logrotate in debug mode does nothing except printing debug messages!  Consider using verbose mode (-v) instead if this is not what you want.

reading config file /etc/logrotate.conf
including /etc/logrotate.d
Ignoring yum.rpmsave, because of *.rpmsave pattern match
reading config file bootlog
reading config file btmp
reading config file chrony
reading config file cyrus-imapd
reading config file dnf
reading config file dpkg
reading config file fedmsg
reading config file glusterfs
reading config file httpd
reading config file iscsiuiolog
reading config file libvirtd
reading config file libvirtd.qemu
reading config file numad
reading config file ppp
reading config file squid
reading config file tinyproxy
reading config file vsftpd
reading config file wtmp
reading config file yum
error: error opening state file /var/lib/logrotate.status: Permission denied
```

Reply:
True, only tested in debug mode with an non-existent state file...

Do we need an opt-out option (`--skip-state-lock`)?

Reply:
The fix looks good, thanks!

As for `--skip-state-lock`, what should be the use case?

If one needs to enforce rotation while there is a stray process of logrotate, they should be able to either kill the process or replace the state file.

If we want to cover platforms where `flock()` is not available, we should probably consider a compile-time check instead.

The only case that comes to my mind is a file system where `flock()` would fail at run-time.

Is there any other use case that I am overseeing?

Anyway, the option is easy to implement and does not constitute any maintenance or administration burden.  I do not have strong opinion about it...

Reply:
Added the option `--skip-state-lock`.

If `flock(2)` is unavailable, compilation will fail, so packagers/user will stumble over it upfront and can tackle it.

I was considering systems where for whatever reason locking in unsupported at runtime or prohibited. At that time logrotate is compiled/shipped and there would be no easy way (other then re-compiling, what might be difficult or prohibited at that system) to get around.

Reply:
Looks good.  Thanks for the update!